### PR TITLE
Add mustacheRenderW

### LINKED ISSUE #3: Variant of renderMustache which collects errors
Currently, `renderMustache` does not give any indication that substitution errors have occurred (other than a return value of `""`, but I dislike relying on stringly typed programming practices). This has caused confusion at least once: see https://github.com/bos/criterion/issues/127. It would be convenient to have a variation on `renderMustache` that also gives error messages, similarly to how [`checkedSubstitute`](http://hackage.haskell.org/package/mustache-2.2.3/docs/Text-Mustache-Render.html#v:checkedSubstitute) from the `mustache` library does this.

### PR DESCRIPTION
**Pull request by phadej**
Fix #3 

@RyanGlScott, is this something you had in mind?

### CODE REVIEW

File: src/Text/Microstache/Type.hs [hunk 1]
```diff
@@ -126,3 +128,20 @@ instance Exception MustacheException where
...
+  | MustacheDirectlyRenderedValue Key
+    -- ^ A complex value such as an Object or Array was directly rendered into the template
+  deriving (Eq, Show, Typeable, Generic)
+
+#if MIN_VERSION_base(4,8,0)
+instance Exception MustacheWarning where
+  displayException (MustacheVariableNotFound key) = 
+    "Referenced value was not provided, key: " ++ T.unpack (showKey key)
+  displayException (MustacheDirectlyRenderedValue key) = 
+    "Complex value rendered as such, key: " ++ T.unpack (showKey key)
```
Comment (RyanGlScott) [inline src/Text/Microstache/Type.hs]: Can you factor this `displayException` implementation out to a top-level function? It'd be handy to be able to use this on pre-`base-4.8` compilers.

File: src/Text/Microstache/Type.hs [hunk 2]
```diff
@@ -126,3 +128,20 @@ instance Exception MustacheException where
 #else
 instance Exception MustacheException
 #endif
+
+data MustacheWarning
+  = MustacheVariableNotFound Key
+    -- ^ The template contained a variable for which there was no data counterpart in the current context
+  | MustacheDirectlyRenderedValue Key
+    -- ^ A complex value such as an Object or Array was directly rendered into the template
```
Comment (RyanGlScott) [inline src/Text/Microstache/Type.hs]: (Disclaimer: I know sufficiently little about mustache templates that the comment below may not make sense.)

I noticed there are only two possible `MustacheWarnings`, but its counterpart in the `mustache` package ([`SubstitutionError`](http://hackage.haskell.org/package/mustache-2.2.3/docs/Text-Mustache-Render.html#t:SubstitutionError)) has several more constructors. Is there a rationale for which ones are included in `MustacheWarning`?

**Comment by RyanGlScott** [comment]
Also, a meta-question: since this is more-or-less a downstream fork of `stache`, should we try to get this upstreamed in `stache` as well? (I opened a [counterpart issue](https://github.com/stackbuilders/stache/issues/23) for `stache` as well - I'm not sure if that prompted this PR, or you just happened to get around to it at the same time.)

File: src/Text/Microstache/Type.hs (hunk 1 above)
Comment (phadej) [inline src/Text/Microstache/Type.hs]: Good idea. Will add.

**Comment by phadej** [comment]
About `stache`, if @mrkkrp want to use this approach, why so not. Yet `microstache` divergesdfrom `stache` by not throwing the exception in the first place (as spec essentially says "rendering shouldn't fail")

File: src/Text/Microstache/Type.hs [hunk 3]
```diff
@@ -109,20 +113,42 @@ instance NFData PName
...
     -- ^ A referenced value was not provided. The exception provides info
     -- about partial in which the issue happened 'PName' and name of the
     -- missing key 'Key'.
   deriving (Eq, Show, Typeable, Generic)
 
-#if MIN_VERSION_base(4,8,0)
-instance Exception MustacheException where
-  displayException (MustacheParserException e) = show e
-  displayException (MustacheRenderException pname key) =
+{-# DEPRECATED MustacheRenderException "Not thrown anymore, will be removed in the next major version of microstache" #-}
```
Comment (RyanGlScott) [inline src/Text/Microstache/Type.hs]: I'm a bit confused by this deprecation for `MustacheRenderException`. Should it actually be `MustacheException` instead?

Comment (phadej) [inline src/Text/Microstache/Type.hs]: No, just `MustacheRenderException`. `MustacheParserException` is thrown by `compileMustacheFile`

Comment (phadej) [inline src/Text/Microstache/Type.hs]: i.e. it was my mistake not to remove `Render` part in the first version.

**Comment by RyanGlScott** [comment]
OK, I had the chance to try this out on `criterion`. I tried replacing `{{{json}}}` with `{{{jsonx}}}` in the default mustache template to see what would happen, and lo and behold:

```
criterion: warning:
  Referenced value was not provided, key: jsonx
```

Beautiful!

However, there's another problem: when you run `criterion` with the default (unmodified) template on any project, it will emit this warning:

```
criterion: warning:
  Complex value rendered as such, key: json
```

That's due to [these lines](https://github.com/bos/criterion/blob/b1a4718cf40e4e35250bf6a1c33239c7bebc4080/templates/default.tpl#L274-L275):

```javascript
  var reports = {{{json}}}; // Will be substituted with [{"reportAnalysis":...}]
  reports.map(mangulate);
```

Is the a way to redesign this to avoid the warning? Apologies for dumping a bunch of `criterion`-specific questions into this pull request, but it seemed like a fitting place to ask.

**Comment by phadej** [comment]
Hmm. The warning about `{{{json}}}` is kind-of correct. I'd say that usually such thing would be suspicious. I'd recommend to `filter (/= MustacheDirectlyRenderedValue (Key ["json"]))` as it's an expected warning.

**Comment by RyanGlScott** [comment]
> I'd recommend to `filter (/= MustacheDirectlyRenderedValue (Key ["json"]))` as it's an expected warning.

Well, that does work, I suppose ;)

I'd say this LGTM. Thanks!

File: src/Text/Microstache/Type.hs (hunk 2 above)
Comment (phadej) [inline src/Text/Microstache/Type.hs]: @RyanGlScott , I'm happy to change an API so adding new warnings would be non-breaking change,
I wonder is someone would need exhaustive matching on warnings (as that will break any time one adds a warning).

With product types it's easy: smart constructors and field accessors/lens. is there good ways for sums?

---

- I found a place where to check for `PartialNotFound`, I'l try to write a test possibly triggering it.
- `SectionTargetNotFound` is subsumed by `MustacheVariableNotFound`
    - I can split it, if that's a good idea?
- `InvalidImplicitSectionContextType` is when

```haskell
>>> let Right templ = compileMustacheText "" "{{#string}} {{.}} {{#.}} ?  {{/.}} {{/string}}"
>>> renderMustache templ $ object [ "string" .= ("foobar" :: Text) ]
" foobar  ?   "
```

```
% cat ex1.json 
{ "string": "foobar" }
% cat ex.mustache 
{{#string}} {{.}} {{#.}} ?  {{/.}} {{/string}}
% ./node_modules/.bin/mustache ex1.json ex.mustache
 foobar  ?   
```

when given `{ string: "string" }`.

Quite obscure, I'm not really sure that something I'd bother to check.

For something like auto-flattening?

```
>>> let Right templ = compileMustacheText "" "{{#foos}}{{#.}}{{#.}}{{.}};{{/.}}{{/.}}{{/foos}}"
>>> renderMustache templ $ object [ "foos" .= ["foo1" :: Text, "foo2"] ]
"foo1;foo2;"
>>> renderMustache templ $ object [ "foos" .= [["foo1" :: Text, "foo2"], ["foo3"]] ]
"foo1;foo2;foo3;"
```

In the first case `mustache` would error, as inner `{{#.}}` is pointing to json string.

**Comment by phadej** [comment]
https://hackage.haskell.org/package/microstache-1.0.1 released

**Comment by mrkkrp** [comment]
> Yet microstache diverged from stache by not throwing the exception in the first place (as spec essentially says "rendering shouldn't fail")

`stache` won't throw in version 1.0.0. We'll be again compatible with the spec. Version 1.0.0 will also use Megaparsec 6, so release of `stache-1.0.0` is pending release of Megaparsec 6.

I'll comment on stackbuilders/stache#23 as well.

**Pull request merged**